File: src/dotnet/tests/Ansight.UnitTests/PurchaseDiagnosticsTests.cs
namespace Ansight.UnitTests;

using Ansight.Purchases;
using Ansight.Tools;
using System.Text.Json;
using System.Text.Json.Nodes;
using Xunit;

public class PurchaseDiagnosticsTests
{
    [Fact]
    public async Task UnknownAndProductionCannotUseAnyEntryPoint()
    {
        var diagnostics = new PurchaseDiagnostics(() => false);
        Assert.False(diagnostics.IsInteropAllowed);
        Assert.Throws<PurchaseEnvironmentException>(() => diagnostics.Record(new("item", "app", 100, Environment: "sandbox")));
        Assert.Throws<PurchaseEnvironmentException>(() => diagnostics.RecordProduct(new("item", true, 100)));
        Assert.Throws<PurchaseEnvironmentException>(() => diagnostics.CreateTransactionReference("token"));
        Assert.Throws<PurchaseEnvironmentException>(() => diagnostics.ImportStoreKit([], [], diagnostics.CaptureGeneration()));
        Assert.Throws<PurchaseEnvironmentException>(() => PurchaseGooglePlay.Record([], 1, false, diagnostics: diagnostics));
        await Assert.ThrowsAsync<PurchaseEnvironmentException>(() => PurchaseStoreKit.RefreshAsync([], diagnostics));
        foreach (var tool in PurchaseTools.Create(diagnostics)) {
            Assert.Throws<PurchaseEnvironmentException>(() => diagnostics.Execute(tool.Id));
            var availability = await tool.GetAvailabilityAsync(new(null, null));
            var result = await ((IJsonTool)tool).ExecuteAsync(new(new(), new("test", null, null)), CancellationToken.None);
            Assert.False(result.IsSuccess);
            Assert.Equal(PurchaseEnvironmentException.ErrorCode, result.ErrorCode);
            Assert.False(availability.IsAvailable);
            Assert.Equal(PurchaseEnvironmentException.ErrorCode, availability.ReasonCode);
        }
        diagnostics.Clear(); // Local cleanup must remain safe even when unavailable.
        Assert.False(new PurchaseDiagnostics().IsInteropAllowed); // Plain net9.0 cannot establish a store sandbox.
    }

    [Fact]
    public void EnvironmentIsRecheckedAndProductionEvidenceIsRejected()
    {
        var allowed = true;
        var diagnostics = new PurchaseDiagnostics(() => allowed);
        diagnostics.Record(new("item", "store", 100, Environment: "sandbox"));
        allowed = false;
        Assert.Throws<PurchaseEnvironmentException>(() => diagnostics.Execute("purchases.get_state"));
        allowed = true;
        Assert.Empty(diagnostics.Execute("purchases.get_state")["observations"]!.AsArray());
        Assert.Throws<PurchaseEnvironmentException>(() => diagnostics.Record(new("item", "store", 100, Environment: "production")));
        Assert.Throws<PurchaseEnvironmentException>(() => diagnostics.ImportStoreKit(
            [new("sandbox", "store", 100, Environment: "sandbox"), new("production", "store", 100, Environment: "production")], [], diagnostics.CaptureGeneration()));
        Assert.Empty(diagnostics.Execute("purchases.get_state")["observations"]!.AsArray());
        Assert.False(new PurchaseDiagnostics(() => throw new Exception()).IsInteropAllowed);
        Assert.True(PurchaseEnvironment.IsAndroidEmulator("ranchu"));
        Assert.True(PurchaseEnvironment.IsAndroidEmulator("goldfish"));
        foreach (var hardware in new string?[] { null, "", "unknown", "test-keys", "generic", "qcom" })
            Assert.False(PurchaseEnvironment.IsAndroidEmulator(hardware));
    }

    [Fact]
    public void ReferencesAreOpaqueAndClearInvalidatesAnInflightImport()
    {
        var diagnostics = new PurchaseDiagnostics(() => true);
        var reference = diagnostics.CreateTransactionReference("raw-purchase-token");
        Assert.Equal(64, reference.Length);
        Assert.Equal(reference, diagnostics.CreateTransactionReference("raw-purchase-token"));
        Assert.NotEqual(reference, new PurchaseDiagnostics(() => true).CreateTransactionReference("raw-purchase-token"));
        var generation = diagnostics.CaptureGeneration();
        diagnostics.Clear();
        Assert.NotEqual(reference, diagnostics.CreateTransactionReference("raw-purchase-token"));
        Assert.Throws<OperationCanceledException>(() => diagnostics.ImportStoreKit([new("item", "store", 100)], [], generation));
        Assert.Empty(diagnostics.Execute("purchases.get_state")["observations"]!.AsArray());
    }

    [Fact]
    public void SharedValidationCases()
    {
        var fixture = JsonNode.Parse(File.ReadAllText(Path.Combine(AppContext.BaseDirectory, "purchase-validation-cases.json")))!;
        var options = new JsonSerializerOptions(JsonSerializerDefaults.Web);
        foreach (var test in fixture["cases"]!.AsArray()) {
            var diagnostics = new PurchaseDiagnostics(() => true);
            foreach (var row in test!["observations"]!.AsArray()) diagnostics.Record(row!.Deserialize<PurchaseObservation>(options)!);
            var result = diagnostics.Execute("purchases.validate", test["arguments"]!.DeepClone().AsObject(), fixture["now"]!.GetValue<long>());
            Assert.True(ToolSchemaValidator.Validate(PurchaseSchemas.Result, result).IsValid);
            Assert.True(result["status"]!.GetValue<string>() == test["status"]!.GetValue<string>(), test["name"]!.GetValue<string>() + ": " + result);
        }
    }

    [Fact]
    public void RetentionAndCursorGapsAreExplicit()
    {
        var diagnostics = new PurchaseDiagnostics(() => true);
        for (var index = 0; index < 300; index++) diagnostics.Record(new("item", "app", index));
        var result = diagnostics.Execute("purchases.get_events", new() { ["after"] = 0 });
        Assert.Equal(256, result["observations"]!.AsArray().Count);
        Assert.Equal(44, result["droppedEvents"]!.GetValue<long>());
        Assert.True(result["cursorGap"]!.GetValue<bool>());
        Assert.Empty(diagnostics.Execute("purchases.get_events", new() { ["after"] = 300 })["observations"]!.AsArray());
        diagnostics.Clear();
        Assert.Throws<ArgumentException>(() => diagnostics.Execute("purchases.get_events", new() { ["after"] = 300 }));
    }

    [Fact]
    public void AllToolsAreReadOnlyAndReadsDoNotMutateEvidence()
    {
        var diagnostics = new PurchaseDiagnostics(() => true);
        diagnostics.Record(new("item", "app", 100, DeliveryCount: 1));
        var before = diagnostics.Execute("purchases.get_state", now: 100).ToJsonString();
        foreach (var tool in PurchaseTools.Create(diagnostics)) Assert.Equal(ToolPolicy.Read, tool.Policy);
        diagnostics.Execute("purchases.validate", new() { ["productId"] = "item", ["expectedEntitled"] = true }, 100);
        Assert.Equal(before, diagnostics.Execute("purchases.get_state", now: 100).ToJsonString());
    }

    [Fact]
    public void MalformedInputsAndUndeclaredArgumentsAreRejected()
    {
        var diagnostics = new PurchaseDiagnostics(() => true);
        Assert.Throws<ArgumentException>(() => diagnostics.Record(new("", "app", 0)));
        Assert.Throws<ArgumentException>(() => diagnostics.Record(new("item", "arbitrary", 0)));
        Assert.Throws<ArgumentException>(() => diagnostics.Execute("purchases.validate", new() { ["productId"] = "item", ["expectedEntitled"] = true, ["receipt"] = "secret" }));
        Assert.Throws<ArgumentException>(() => diagnostics.Execute("purchases.validate", new() { ["productId"] = "item", ["expectedEntitled"] = true, ["maxAgeMilliseconds"] = 0 }));
    }
}

File: src/ios/Tests/AnsightCoreTests/PurchaseDiagnosticsTests.swift
import Foundation
import XCTest
@testable import AnsightCore

final class PurchaseDiagnosticsTests: XCTestCase {
    func testEnvironmentGuardCoversEveryEntryPoint() async throws {
        let diagnostics = PurchaseDiagnostics(environmentAllowed: { false })
        XCTAssertFalse(diagnostics.isInteropAllowed)
        var row = PurchaseObservation(productId: "item", source: "app")
        row.environment = "sandbox"
        XCTAssertThrowsError(try diagnostics.record(row))
        XCTAssertThrowsError(try diagnostics.recordProduct(PurchaseProduct(productId: "item", available: true)))
        XCTAssertThrowsError(try diagnostics.createTransactionReference("token"))
        XCTAssertThrowsError(try diagnostics.recordStoreKitBatch([], products: [], generation: diagnostics.captureGeneration()))
        for action in ["record", "recordProduct", "reference", "register", "clear", "refreshStoreKit"] + PurchaseDiagnostics.toolIds {
            XCTAssertThrowsError(try diagnostics.command("{\"action\":\"\(action)\"}")) { error in
                XCTAssertEqual(error as? PurchaseDiagnosticsError, .environmentNotAllowed)
            }
            do { _ = try await diagnostics.commandAsync("{\"action\":\"\(action)\"}"); XCTFail("Allowed blocked bridge") }
            catch { XCTAssertEqual(error as? PurchaseDiagnosticsError, .environmentNotAllowed) }
        }
        for tool in diagnostics.tools() {
            XCTAssertThrowsError(try diagnostics.execute(tool.descriptor.id))
            let availability = tool.availability(context: .init(sessionId: nil, requestId: nil))
            let result = try (tool as! any AnsightJSONTool).execute(arguments: [:])
            XCTAssertEqual(result.errorCode, PurchaseDiagnosticsError.environmentErrorCode)
            XCTAssertFalse(availability.available)
            XCTAssertEqual(availability.reasonCode, PurchaseDiagnosticsError.environmentErrorCode)
        }
        if #available(macOS 12.0, iOS 15.0, *) {
            do { try await PurchaseStoreKit.refresh(productIds: [], diagnostics: diagnostics); XCTFail("Allowed blocked StoreKit") }
            catch { XCTAssertEqual(error as? PurchaseDiagnosticsError, .environmentNotAllowed) }
        }
        diagnostics.clear()
    }

    func testGuardRechecksAndRejectsProductionEvidence() throws {
        var allowed = true
        let diagnostics = PurchaseDiagnostics(environmentAllowed: { allowed })
        var row = PurchaseObservation(productId: "item", source: "store")
        row.environment = "sandbox"
        try diagnostics.record(row)
        allowed = false
        XCTAssertThrowsError(try diagnostics.execute("purchases.get_state"))
        allowed = true
        guard case .object(let result) = try diagnostics.execute("purchases.get_state") else { return XCTFail() }
        XCTAssertEqual(result["observations"], .array([]))
        row.environment = "production"
        XCTAssertThrowsError(try diagnostics.record(row))
        XCTAssertThrowsError(try diagnostics.recordStoreKitBatch([PurchaseObservation(productId: "safe", source: "store"), row], products: [], generation: diagnostics.captureGeneration()))
        #if targetEnvironment(simulator)
        XCTAssertTrue(PurchaseEnvironment.isAllowed())
        #else
        XCTAssertFalse(PurchaseEnvironment.isAllowed())
        XCTAssertFalse(PurchaseDiagnostics().isInteropAllowed)
        #endif
    }

    func testReferenceIsolationAndInflightRefreshAfterClear() throws {
        let diagnostics = PurchaseDiagnostics(environmentAllowed: { true })
        let reference = try diagnostics.createTransactionReference("raw-purchase-token")
        XCTAssertEqual(reference.count, 64)
        XCTAssertEqual(reference, try diagnostics.createTransactionReference("raw-purchase-token"))
        XCTAssertNotEqual(reference, try PurchaseDiagnostics(environmentAllowed: { true }).createTransactionReference("raw-purchase-token"))
        let generation = diagnostics.captureGeneration()
        diagnostics.clear()
        XCTAssertNotEqual(reference, try diagnostics.createTransactionReference("raw-purchase-token"))
        XCTAssertThrowsError(try diagnostics.recordStoreKitBatch([PurchaseObservation(productId: "item", source: "store")], products: [], generation: generation))
    }

    func testSharedValidationCases() throws {
        let fixtureURL = URL(fileURLWithPath: #filePath).deletingLastPathComponent().deletingLastPathComponent().deletingLastPathComponent().deletingLastPathComponent().deletingLastPathComponent().appendingPathComponent("docs/contracts/purchases/validation-cases.json")
        let fixture = try JSONSerialization.jsonObject(with: Data(contentsOf: fixtureURL)) as! [String: Any]
        for test in fixture["cases"] as! [[String: Any]] {
            let diagnostics = PurchaseDiagnostics(environmentAllowed: { true })
            for row in test["observations"] as! [[String: Any]] {
                let request = try JSONSerialization.data(withJSONObject: ["action": "record", "observation": row])
                _ = try diagnostics.command(String(decoding: request, as: UTF8.self))
            }
            let arguments = try JSONDecoder().decode([String: JSONValue].self, from: JSONSerialization.data(withJSONObject: test["arguments"]!))
            let result = try diagnostics.execute("purchases.validate", arguments: arguments, now: 100000)
            XCTAssertTrue(AnsightToolSchemaValidator.validate(schema: PurchaseSchemas.result, value: result).isEmpty)
            guard case .object(let value) = result else { return XCTFail("Invalid report") }
            XCTAssertEqual(value["status"], .string(test["status"] as! String), test["name"] as! String)
        }
    }
    func testRetentionAndPassiveReads() throws {
        let diagnostics = PurchaseDiagnostics(environmentAllowed: { true })
        for value in 0..<300 { try diagnostics.record(PurchaseObservation(productId: "item", source: "app", observedAt: Int64(value))) }
        guard case .object(let result) = try diagnostics.execute("purchases.get_events") else { return XCTFail() }
        XCTAssertEqual(result["droppedEvents"], .integer(44))
        XCTAssertEqual(result["cursorGap"], .bool(true))
        XCTAssertEqual(result["nextCursor"], .integer(300))
        for tool in diagnostics.tools() { XCTAssertEqual(tool.descriptor.policy, .read) }
        let before = try diagnostics.execute("purchases.get_state", now: 300)
        _ = try diagnostics.execute("purchases.validate", arguments: ["productId": .string("item"), "expectedEntitled": .bool(true)], now: 300)
        XCTAssertEqual(try diagnostics.execute("purchases.get_state", now: 300), before)
        diagnostics.clear()
        XCTAssertThrowsError(try diagnostics.execute("purchases.get_events", arguments: ["after": .integer(300)]))
    }
}
